prov/usnic/src/usdf_wait.c

```diff
@@ -0,0 +1,338 @@
+		return -FI_EOPNOTSUPP;
+	}
+
+	return usd_close(dev);
+}
+
+/* Non-static because this is exported due to being returned as a callback for
+ * fabric ops.
+ *
+ * Supporting wait objects in the usNIC provider is done using an epoll
+ * context. When fi_wait_open is called an epoll context is created using
+ * epoll_create1. This simplifies multi-CQ support and also allows us to get
+ * around a limitation of the usNIC provider. IB completion channels are opened
+ * on the domain because we have a context associated with the domain. At
+ * fi_wait_open time, we only have access to the fabric. It isn't guaranteed
+ * that a domain has been opened yet. The epoll context approach allows us to
+ * defer creating the completion channel for the CQ until CQ open time.
+ */
+int usdf_wait_open(struct fid_fabric *fabric, struct fi_wait_attr *attr,
+		struct fid_wait **waitset)
+{
+	struct usdf_wait *wait_priv;
+	struct usdf_fabric *fabric_priv;
+	int epfd;
+	int ret;
+
+	USDF_TRACE_SYS(FABRIC, "\n");
+
+	switch (attr->wait_obj) {
+	case FI_WAIT_UNSPEC:
+	case FI_WAIT_FD:
+		break;
+	default:
+		USDF_WARN_SYS(FABRIC, "unsupported wait object type\n");
+		ret = -FI_EINVAL;
+		goto error;
+	}
+
+	fabric_priv = fab_fidtou(fabric);
+	ret = usdf_wait_check_support(fabric_priv);
+	if (ret)
+		goto error;
+
+	epfd = epoll_create1(0);
+	if (epfd < 0) {
+		USDF_WARN_SYS(FABRIC, "failed to create epoll fd[%d]\n", errno);
+		ret = -errno;
+		goto error;
+	}
+
+	USDF_DBG_SYS(FABRIC, "successfully created epoll fd: %d\n", epfd);
+
+	wait_priv = calloc(1, sizeof(*wait_priv));
+	if (!wait_priv) {
+		USDF_WARN_SYS(FABRIC,
+				"unable to allocate memory for usdf_wait obj");
+		ret = -FI_ENOMEM;
+		goto calloc_fail;
+	}
+
+	wait_priv->wait_fid.fid.fclass = FI_CLASS_WAIT;
+	wait_priv->wait_fid.fid.ops = &usdf_wait_fi_ops;
+	wait_priv->wait_fid.ops = &usdf_wait_ops;
+	wait_priv->wait_fid.fid.context = 0;
+	wait_priv->wait_fabric = fabric_priv;
+	wait_priv->wait_obj = attr->wait_obj;
+	wait_priv->object.epfd = epfd;
+
+	atomic_initialize(&wait_priv->wait_refcnt, 0);
+	fastlock_init(&wait_priv->lock);
+	dlist_init(&wait_priv->list);
+
+	atomic_inc(&wait_priv->wait_fabric->fab_refcnt);
+
+	*waitset = &wait_priv->wait_fid;
+
+	return FI_SUCCESS;
+
+calloc_fail:
+	close(epfd);
+error:
+	*waitset = NULL;
+	return ret;
+}
+
+/* Close a wait object. Make sure all resources associated with the wait object
+ * have been closed.
+ */
+static int usdf_wait_close(struct fid *waitset)
+{
+	struct usdf_wait *wait_priv;
+
+	USDF_TRACE_SYS(FABRIC, "\n");
+	if (!waitset) {
+		USDF_WARN_SYS(FABRIC, "invalid input.\n");
+		return -FI_EINVAL;
+	}
+
+	wait_priv = wait_ftou(waitset);
+
+	if (atomic_get(&wait_priv->wait_refcnt) > 0) {
+		USDF_DBG_SYS(FABRIC,
+				"failed to close waitset with non-zero refcnt");
+		return -FI_EBUSY;
+	}
+
+	switch (wait_priv->wait_obj) {
+	case FI_WAIT_UNSPEC:
+	case FI_WAIT_FD:
+		close(wait_priv->object.epfd);
+		break;
+	default:
+		USDF_WARN_SYS(FABRIC,
```

Comment:
> Shouldn't this check-the-type-of-the-waitset be at the top of the routine? I.e., if it's an unknown waitset type, we've already gotten its refcount and done things to it that may be erroneous.

Actually, just looked at this more in depth and it's occurring in the correct order. That get doesn't do anything other than get the current value and make sure it's positive.